utils/residual_stream.py

```diff
@@ -0,0 +1,107 @@
+from typing import List, Tuple
+from jaxtyping import Int, Float, Bool
+from typeguard import typechecked
+import torch
+from torch import Tensor
+from transformer_lens import HookedTransformer
+from transformer_lens.utils import get_act_name
+from utils.prompts import get_dataset, PromptType
+
+
+def get_resid_name(layer: int, model: HookedTransformer) -> Tuple[str, int]:
+    resid_type = 'resid_pre'
+    if layer == model.cfg.n_layers:
+        resid_type = 'resid_post'
+        layer -= 1
+    return get_act_name(resid_type, layer), layer
+
+
+class ResidualStreamDataset:
+
+    @typechecked
+    def __init__(
+        self, 
+        prompt_strings: List[str], 
+        prompt_tokens: Int[Tensor, "batch pos"], 
+        binary_labels: Bool[Tensor, "batch"],
+        model: HookedTransformer,
+        prompt_type: PromptType,
+    ) -> None:
+        assert len(prompt_strings) == len(prompt_tokens)
+        assert len(prompt_strings) == len(binary_labels)
+        self.prompt_strings = prompt_strings
+        self.prompt_tokens = prompt_tokens
+        self._binary_labels = binary_labels
+        self.model = model
+        self.prompt_type = prompt_type
+        example_str_tokens = model.to_str_tokens(prompt_strings[0])
+        self.example = [f"{i}:{tok}" for i, tok in enumerate(example_str_tokens)]
+        self.placeholder_dict = prompt_type.get_placeholder_positions(example_str_tokens)
+        label_positions = [pos for _, positions in self.placeholder_dict.items() for pos in positions]
+        self.str_labels = [
+            ''.join([model.to_str_tokens(prompt)[pos] for pos in label_positions]) 
+            for prompt in prompt_strings
+        ]
+
+    @property
+    def binary_labels(self) -> Bool[Tensor, "batch"]:
+        return self._binary_labels.cpu().detach()
+
+    def __len__(self) -> int:
+        return len(self.prompt_strings)
+    
+    def __eq__(self, other: object) -> bool:
+        return set(self.prompt_strings) == set(other.prompt_strings)
+    
+    @typechecked
+    def embed(self, position_type: str, layer: int) -> Float[Tensor, "batch d_model"]:
+        """
+        Returns a dataset of embeddings at the specified position and layer.
```

Comment:
@curt-tigges I added a few comments like this to the function doctstrings. Are these in the right sort of style?